File: app/src/main/java/com/salino/sali/data/service/CategoryDetectionCoordinator.kt
package com.salino.sali.data.service

import com.salino.sali.data.local.CategoryClassificationStore
import com.salino.sali.data.model.ItemCategory
import com.salino.sali.domain.service.AiCategoryClassifier
import com.salino.sali.domain.service.CategoryAutoDetector
import com.salino.sali.util.normalizeItemName
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Two-stage classification: strict local keywords first, Gemini only when keywords abstain.
 * Local layer never uses fuzzy/substring guessing — null means "not sure, ask AI".
 */
@Singleton
class CategoryDetectionCoordinator @Inject constructor(
    private val keywordDetector: CategoryAutoDetector,
    private val aiClassifier: AiCategoryClassifier,
    private val cache: CategoryClassificationStore
) {
    fun detectWithKeywords(itemName: String): ItemCategory? =
        keywordDetector.detectCategory(itemName)

    suspend fun detectWithAi(itemName: String): ItemCategory? {
        val normalized = normalizeItemName(itemName)
        if (normalized.length < 2) return null

        cache.get(normalized)?.let { return it }

        val category = aiClassifier.classify(itemName) ?: return null
        cache.put(normalized, category)
        return category
    }

    fun isAutoDetectable(category: ItemCategory): Boolean =
        category != ItemCategory.OTHER
}

File: mobile/src/services/categoryDetectionCoordinator.ts
import { ItemCategory } from '../models';
import { normalizeItemName } from '../utils/textUtils';
import { detectCategory } from './categoryDetector';
import { classifyWithAi } from './firebaseAiCategoryClassifier';
import { categoryClassificationCache } from '../local/categoryClassificationCache';

/**
 * Two-stage classification: strict local keywords first, Gemini only when
 * keywords abstain.  Mirrors Android `CategoryDetectionCoordinator`.
 */
export function detectWithKeywords(itemName: string): ItemCategory | null {
  return detectCategory(itemName);
}

export async function detectWithAi(itemName: string): Promise<ItemCategory | null> {
  const normalized = normalizeItemName(itemName);
  if (normalized.length < 2) return null;

  const cached = await categoryClassificationCache.get(normalized);
  if (cached) return cached;

  const category = await classifyWithAi(itemName);
  if (!category) return null;

  await categoryClassificationCache.put(normalized, category);
  return category;
}

/** Auto-detected badge should only show for real categories, not OTHER. */
export function isAutoDetectable(category: ItemCategory): boolean {
  return category !== ItemCategory.OTHER;
}
